%% start of file `template.tex'.
%% Copyright 2006-2010 Xavier Danaux (xdanaux@gmail.com).
%% Copyright 2010-2011 Mark Liu (markwayneliu@gmail.com).
%
% This work may be distributed and/or modified under the
% conditions of the LaTeX Project Public License version 1.3c,
% available at http://www.latex-project.org/lppl/.

\documentclass[11pt,a4paper,sans]{moderncv}

\usepackage{verbatim}

% moderncv themes
\moderncvstyle{classic}
\moderncvcolor{burgundy}

% character encoding
\usepackage[utf8]{inputenc}                   % replace by the encoding you are using

% adjust the page margins
\usepackage[scale=0.8]{geometry}
%\setlength{\hintscolumnwidth}{3cm}						% if you want to change the width of the column with the dates
%\AtBeginDocument{\setlength{\maketitlenamewidth}{6cm}}  % only for the classic theme, if you want to change the width of your name placeholder (to leave more space for your address details
%\AtBeginDocument{\recomputelengths}                     % required when changes are made to page layout lengths


% personal data
\firstname{Thomas A.}
\familyname{Sasani}
\email{thomas.a.sasani@gmail.com}                      % optional, remove the line if not wanted
\homepage{http://tomsasani.github.io}
\extrainfo{GitHub: \Colorhref{https://github.com/tomsasani}{@tomsasani}}
%\extrainfo{Twitter: \Colorhref{https://twitter.com/tomsasani}{@tomsasani}}
%\extrainfo{\url\Colorhref{https://scholar.google.com/citations?user=T_Tq0UYAAAAJ&hl=en}{Google Scholar}}                % optional, remove the line if not wanted
%\extrainfo{\url{http://markliu.me}} % optional, remove the line if not wanted

% to show numerical labels in the bibliography; only useful if you make citations in your resume
%\makeatletter
%\renewcommand*{\bibliographyitemlabel}{\@biblabel{\arabic{enumiv}}}
%\makeatother

%\nopagenumbers{}                             % uncomment to suppress automatic page numbering for CVs longer than one page
%----------------------------------------------------------------------------------
%            content
%----------------------------------------------------------------------------------
\begin{document}
\definecolor{color1}{rgb}{0.22,0.45,0.70}
\newcommand\Colorhref[3][color1]{\href{#2}{\small\color{#1}#3}}
\maketitle



\section{Education}
\cventry{2019}{Ph.D, Human Genetics}{University of Utah}{Salt Lake City, UT}{}{}
%\cvline{advisor:}{\small Professor Aaron Quinlan}
\cventry{2015}{B.A, Biochemistry}{Lawrence University}{Appleton, WI}{\emph{summa cum laude}}{}
%\cvline{advisor:}{\small Professor Brian Piasecki}

\section{Experience}

\cventry{10/22-pres.}{Staff Research Scientist}{Quinlan Lab}{Univ. of Utah}{Dept. of Human Genetics}{
\begin{itemize}
    \item Developing new methods to analyze germline mutation and genome evolution 
\end{itemize}
}

\cventry{5/21-10/22}{Senior Data Scientist}{Recursion Pharmaceuticals}{}{}{
\begin{itemize}
\item Developed new computational methods to analyze data from massive cellular imaging experiments
\end{itemize}
}

\cventry{3/20-5/21}{Postdoctoral Fellow}{Harris Lab}{Univ. of Washington}{Dept. of Genome Sciences}{
\begin{itemize}
\item Discovered alleles that influence the germline mutation rate using whole-genome sequencing data
\item Analyzed single-cell sequencing data to characterize the mutational landscape of spermatogenesis
\end{itemize}
}

\cventry{4/16-3/20}{Graduate Research Assistant}{Quinlan Lab}{Univ. of Utah}{Dept. of Human Genetics}{
\begin{itemize}
\item Analyzed whole-genome sequencing data from large multi-generational families to identify
post-zygotic mosaicism and variability in human germline mutation rates
\item Used the Oxford Nanopore Technologies platform to sequence DNA virus genomes under strong selective pressure during experimental evolution
\end{itemize}
}

\section{Skills}{}
\cventry{Programming}{Python (proficient), R (familiar), SQL (familiar)}{}{}{}{}
\cventry{Computing}{Unix, git, Sun Grid Engine, Amazon EC2}{}{}{}{}
\cventry{Data analysis}{numpy, scipy, sklearn, pandas, jupyter, unit testing frameworks (pytest)}{}{}{}{}
\cventry{Visualization}{matplotlib, ggplot2, plotly + Dash}{}{}{}{}

\section{Selected Academic Publications \Colorhref{https://scholar.google.com/citations?user=T_Tq0UYAAAAJ&hl=en}{(full list at \underline{Google Scholar})}}

\cvline{2023}{\textbf{Sasani TA}, Quinlan AR, Harris KE. Epistasis between mutator alleles contributes to germline mutation rate variability in laboratory mice. \Colorhref{https://www.biorxiv.org/content/10.1101/2023.04.25.537217v1}{\emph{bioRxiv}}. \Colorhref{https://github.com/quinlan-lab/proj-mutator-mapping}{\textbf{Code.}}}

\cvline{2022}{\textbf{Sasani TA}, Ashbrook DG, Beichman AC, Lu L, Palmer AA, Williams RW, Pritchard JK, Harris K. A natural mutator allele shapes mutation spectrum variation in mice. \Colorhref{https://www.nature.com/articles/s41586-022-04701-5}{\emph{Nature}}. \Colorhref{https://github.com/tomsasani/bxd_mutator_manuscript}{\textbf{Code.}}}

\cvline{2019}{\textbf{Sasani TA}, Pedersen BS, Gao Z, Baird L, Przeworski M, Quinlan AR, Jorde LB. Large, three-generation human families reveal post-zygotic mosaicism and variability in germline mutation accumulation. \Colorhref{https://elifesciences.org/articles/46922}{\emph{eLife}}. \Colorhref{https://github.com/quinlan-lab/ceph-dnm-manuscript}{\textbf{Code.}}
\Colorhref{https://www.thenakedscientists.com/articles/interviews/how-many-mutations-do-parents-pass}{Interview on the Naked Scientists podcast.}}

\section{Funding and Awards}{}
\cventry{2020 - 2021}{NIH T32 Postdoctoral Genome Sciences Training Grant}{}{}{}{}
\cventry{2017 - 2019}{NIH T32 Predoctoral Genetics Training Grant}{}{}{}{}
\cventry{2017}{Lassonde Student Innovator}{University of Utah}{}{}{}
%\cventry{2016 \& 2017}{Epstein Award Semifinalist}{American Society of Human Genetics}{}{}{}

\end{document}
